src/ray/raylet/node_manager.cc

```diff
@@ -212,22 +212,23 @@ ray::Status NodeManager::RegisterGcs() {
 }
 
 void NodeManager::KillWorker(std::shared_ptr<Worker> worker) {
+#ifdef _WIN32
+  // TODO(mehrdadn): Implement implement graceful process termination mechanism
+#else
   // If we're just cleaning up a single worker, allow it some time to clean
   // up its state before force killing. The client socket will be closed
   // and the worker struct will be freed after the timeout.
-  kill(worker->Pid(), SIGTERM);
+  kill(worker->Process()->id(), SIGTERM);
+#endif
 
   auto retry_timer = std::make_shared<boost::asio::deadline_timer>(io_service_);
   auto retry_duration = boost::posix_time::milliseconds(
       RayConfig::instance().kill_worker_timeout_milliseconds());
   retry_timer->expires_from_now(retry_duration);
   retry_timer->async_wait([retry_timer, worker](const boost::system::error_code &error) {
-    RAY_LOG(DEBUG) << "Send SIGKILL to worker, pid=" << worker->Pid();
-    // Force kill worker. TODO(mehrdadn, rkn): The worker may have already died
-    // and had its PID reassigned to a different process, at least on Windows.
-    // On Linux, this may or may not be the case, depending on e.g. whether
-    // the process has been already waited on. Regardless, this must be fixed.
-    kill(worker->Pid(), SIGKILL);
+    RAY_LOG(DEBUG) << "Send SIGKILL to worker, pid=" << worker->Process()->id();
+    // Force kill worker
+    worker->Process()->terminate();
```

Comment:
Sounds like we shouldn't wait for the child process' death here?

Reply:
It waits internally :)

https://github.com/boostorg/process/blob/6a4d2ff72114ef47c7afaf92e1042aca3dfa41b0/include/boost/process/detail/posix/terminate.hpp#L24-L30

Reply:
Yes I understand that part:) What I meant is that we probably shouldn't wait for child's death here, since it would potentially block raylet's event loop to process other events, as raylet is single-threaded.

Reply:
Oh I see. Actually, I just noticed the wait uses `WNOHANG`, so it shouldn't block. :) But that probably means I need to explicitly wait for it [here](/ray-project/ray/pull/6510/files/bfcfb57939fa266def58c769b7e4e5ad7c54cd28#diff-64c50c0cc29e8f9f96ebf5df2b668f1cR107).

Reply:
Oh shoot, actually it didn't use `WNOHANG` in our version, that's why I didn't notice it when I looked at it locally! [They added `WNOHANG` in Boost 1.72.](https://github.com/boostorg/process/commit/28126b3432433c025606a84474c7afb5dec88daf#diff-2c76d877b8c53a51eddc9a558a2eb980R30) So I would say let's keep this for now and see if it's an issue? Then maybe upgrade if it poses a problem. (CC @edoakes, any thoughts?)

Reply:
Or actually, I'll add a patch for it myself :)

Reply:
Just added the patch and fixed the other issues. Thanks for the feedback! Let me know if you see any other potential issues.

Reply:
thanks.